### PR TITLE
TS-4341

### PR DESCRIPTION
**Pull request by jacksontj**
TS-4341

### CODE REVIEW

**Comment by jacksontj** [comment]
This isn't complete yet, still need to add handling for timeouts etc., but I wanted to get the PR opened to get some eyes on it :)

**Comment by bgaff** [comment]
:eyes: I'm on it.

**Comment by jacksontj** [comment]
Well, it seems that this doesn't leak in the queue-- but I did find another bug (https://issues.apache.org/jira/browse/TS-4343) which we'll probably want to fix at the same time as this :/

**Comment by jacksontj** [comment]
This PR now includes the patches from https://github.com/apache/trafficserver/pull/562 and https://github.com/apache/trafficserver/pull/569  as they are related (and interdependent)

**Comment by jacksontj** [comment]
@bgaff @jpeach @zwoop ready for review :)

File: lib/ts/ink_inet.cc [hunk 1]
```diff
@@ -307,6 +307,26 @@ ats_ip_hash(sockaddr const *addr)
...
+    unsigned char c[16];
+  } zret;
+
+  zret.i = 0;
+  if (ats_is_ip4(addr)) {
+    zret.i = (static_cast<uint64_t>(ats_ip4_addr_cast(addr)) << 16) | (ats_ip_port_cast(addr));
+  } else if (ats_is_ip6(addr)) {
+    ink_code_md5(const_cast<uint8_t *>(ats_ip_addr8_cast(addr)), TS_IP6_SIZE, zret.c);
+    // now replace the bottom 16bits so we can account for the port.
+    zret.b[3] = ats_ip_port_cast(addr);
```
Comment (jacksontj) [inline lib/ts/ink_inet.cc]: It seems that this returns nothing/garbage in the ipv6 case-- as both lines in ipv6 simply add to `.b` and `.c`, but the method only returns `.i`.

Comment (jacksontj) [inline lib/ts/ink_inet.cc]: That being said, it seems to do the same thing in `ats_ip_hash`, which also seems like it wouldn't work :/

Comment (jpeach) [inline lib/ts/ink_inet.cc]: ``zret`` is a union; it's the same data.

Comment (jacksontj) [inline lib/ts/ink_inet.cc]: @jpeach  thanks for calling out my stupid, for whatever reason I read it as struct ;)

**Comment by bgaff** [comment]
@jacksontj I'm going to merge my pull request, can you please update yours once my code lands? Thanks

**Pull request merged**